File: Lab12/backpack.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);

    int n, x;
    cin >> n >> x;

    vector<int> prices(n), pages(n);

    for (int i = 0; i < n; i++) {
        cin >> prices[i];
    }

    for (int i = 0; i < n; i++) {
        cin >> pages[i];
    }

    vector<int> dp(x + 1, 0);

    for (int i = 0; i < n; i++) {
        for (int w = x; w >= prices[i]; w--) {
            dp[w] = max(dp[w], dp[w - prices[i]] + pages[i]);
        }
    }

    cout << dp[x] << endl;

    return 0;
}

File: Lab12/backpack.py
def book_shop(n: int, max_price: int, prices: list, pages: list) -> int:
    dp = [0] * (max_price + 1)

    for i in range(n):
        for w in range(max_price, prices[i] - 1, -1):
            dp[w] = max(dp[w], dp[w - prices[i]] + pages[i])

    return dp[max_price]


n, x = map(int, input().split())
prices = list(map(int, input().split()))
pages = list(map(int, input().split()))

print(book_shop(n, x, prices, pages))
